internals/overlord/checkstate/handlers.go

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) 2024 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package checkstate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	tombpkg "gopkg.in/tomb.v2"
+
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/overlord/state"
+	"github.com/canonical/pebble/internals/plan"
+)
+
+func (m *CheckManager) doPerformCheck(task *state.Task, tomb *tombpkg.Tomb) error {
+	m.state.Lock()
+	changeID := task.Change().ID()
+	var details checkDetails
+	err := task.Get(checkDetailsAttr, &details)
+	config := m.state.Cached(performConfigKey{changeID}).(*plan.Check)
+	m.state.Unlock()
+	if err != nil {
+		return fmt.Errorf("cannot get check details for perform-check task %q: %v", task.ID(), err)
+	}
+
+	logger.Debugf("Performing check %q with period %v", details.Name, config.Period.Value)
+	ticker := time.NewTicker(config.Period.Value)
+	defer ticker.Stop()
+
+	chk := newChecker(config)
+	for {
+		select {
+		case <-ticker.C:
+			err := runCheck(tomb.Context(nil), chk, config.Timeout.Value)
+			if err != nil {
```

Comment:
This is probably a better way to deal with the tomb dying during or before the check starts.
```
if errors.Is(err, context.Canceled) && tomb.Dying() {
    return tomb.Err()
} else if err != nil {
...
```

Reply:
Good point, done.